grudge/symbolic/operators.py

```diff
@@ -651,20 +658,37 @@ def matrix(self, afgrp, volgrp, dtype):
                     afgrp.nunit_dofs),
                 dtype=dtype)
 
-        from modepy.tools import UNIT_VERTICES
         import modepy as mp
-        for iface, fvi in enumerate(
-                volgrp.mesh_el_group.face_vertex_indices()):
-            face_vertices = UNIT_VERTICES[volgrp.dim][np.array(fvi)].T
-            matrix[:, iface, :] = mp.nodal_face_mass_matrix(
-                    volgrp.basis(), volgrp.unit_nodes, afgrp.unit_nodes,
-                    volgrp.order,
-                    face_vertices)
-
-        # np.set_printoptions(linewidth=200, precision=3)
-        # matrix[np.abs(matrix) < 1e-13] = 0
-        # print(matrix)
-        # 1/0
+        from meshmode.discretization import ElementGroupWithBasis
+
+        vol_basis = volgrp.basis_obj()
+        faces = mp.faces_for_shape(volgrp.shape)
+
+        for iface, face in enumerate(faces):
+            # NOTE: Quadrature degree determined such that
+            # we either guarantee exact integration of degree 2N
+            # polynomials, or use a higher-order quadrature rule
+            # provided by the user (for example, when doing overintegration)
+            face_quad_degree = 2*max(volgrp.order, afgrp.order)
+            face_quadrature = mp.quadrature_for_space(
+                mp.space_for_shape(face, face_quad_degree),
+                face
+            )
+            assert face_quadrature.exact_to >= face_quad_degree
+
+            if isinstance(afgrp, ElementGroupWithBasis):
+                face_basis = afgrp.basis_obj()
+            else:
+                face_basis = mp.basis_for_space(
+                    mp.space_for_shape(face, afgrp.order),
+                    face
+                )
+
+            matrix[:, iface, :] = mp.nodal_mass_matrix_for_face(
+                face, face_quadrature,
+                face_basis.functions, vol_basis.functions,
+                volgrp.unit_nodes, afgrp.unit_nodes,
+            )
```

Comment:
@inducer In light of https://github.com/inducer/modepy/pull/28, does this mean we should _always_ be using `nodal_quad_mass_matrix_for_face`? Or only when we have a quadrature-based element group on the faces?

Reply:
@inducer This doesn't work out of the box. The shape of the guy being returned from `nodal_quad_mass_matrix_for_face` just flat out does not match with that this method is expecting:
```python
matrix = np.empty(
        (volgrp.nunit_dofs,
            nfaces,
            afgrp.nunit_dofs),
        dtype=dtype)
```
And when I pass `afgrp._quadrature_rule()` through the method, I'm getting some einsum value error...
`ValueError: einstein sum subscripts string contains too many subscripts for operand 1`.

I will push my latest commit so you can try and see for yourself (https://github.com/inducer/grudge/pull/62/commits/5c047f6069fa6fb29be097754d04a454e711f564).

Run this to see exactly what I'm seeing:
```
py.test test_grudge.py -k 'test_improvement_quadrature' -x 
```

I will try to spend more time this afternoon digging into what's going on... but I'd appreciate it if you could try it and/or let me know if you see something immediately wrong.

Reply:
> only when we have a quadrature-based element group on the faces?

Only then.

Reply:
If you think about it, they are two different numerical procedures. In the overintegrated case, the quadrature is "exposed". In the nodal case, we use quadrature to compute integrals, but the input data for the matrix is in terms of Lagrange coefficients. For example, the latter needs to know what the trial basis is, the former does not.

Reply:
> but I'd appreciate it if you could try it and/or let me know if you see something immediately wrong.

Needs https://github.com/inducer/meshmode/pull/156